### PR TITLE
DOC + 3d example

### LINKED ISSUE #111: sphinx-gallery: 3d/scattering3d_qm7
Please add basic narrative, display, fix the number, just fix it

Please close this issue only when you're happy with the sphinx-gallery.

### LINKED ISSUE #309: Ill-conditioning in `scattering3d_qm7.py`
When running this, I get
```
...
Ridge regression, alpha: 1.0, MAE: 5.897314548492432, RMSE: 8.19788932800293
/mnt/xfs1/home/janden/local/anaconda3/envs/kymatio_cuda90/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: scipy.linalg.solve
Ill-conditioned matrix detected. Result is not guaranteed to be accurate.
Reciprocal condition number1.719067e-08
  overwrite_a=True).T
/mnt/xfs1/home/janden/local/anaconda3/envs/kymatio_cuda90/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: scipy.linalg.solve
Ill-conditioned matrix detected. Result is not guaranteed to be accurate.
Reciprocal condition number1.847256e-08
  overwrite_a=True).T
/mnt/xfs1/home/janden/local/anaconda3/envs/kymatio_cuda90/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: scipy.linalg.solve
Ill-conditioned matrix detected. Result is not guaranteed to be accurate.
Reciprocal condition number1.704315e-08
  overwrite_a=True).T
/mnt/xfs1/home/janden/local/anaconda3/envs/kymatio_cuda90/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: scipy.linalg.solve
Ill-conditioned matrix detected. Result is not guaranteed to be accurate.
Reciprocal condition number1.990118e-08
  overwrite_a=True).T
/mnt/xfs1/home/janden/local/anaconda3/envs/kymatio_cuda90/lib/python3.7/site-packages/sklearn/linear_model/ridge.py:125: LinAlgWarning: scipy.linalg.solve
Ill-conditioned matrix detected. Result is not guaranteed to be accurate.
Reciprocal condition number1.824673e-08
  overwrite_a=True).T
...
```
Seems like `alpha = 1.0` is a bad idea? Why do we have this in here?

### LINKED ISSUE #421: BUG Working examples with the new frontend
Everything there: https://github.com/kymatio/kymatio/tree/kymatio-dev-frontend/examples must run, frontend must be specified.

- [x] 1D
- [x] 2D
- [x] 3D

### LINKED ISSUE #480: ModuleNotFoundErrors
- [x] 1D
- [x] 2D
- [x] 3D

Hi everyone,

In the current kymatio-v2 branch, there are no __init__.py files in the kymatio.frontend, kymatio.scattering2d.backend and kymatio.scattering2d.core packages. It leads on my side to ModuleNotFoundErrors when trying to run for instance examples/2d/cifar.py after installing kymatio by python setup.py install (or similarly pip install . in the kymatio head folder). It seems to be linked to the find_packages function of setuptools: "find_packages() walks the target directory, filtering by inclusion patterns, and finds Python packages (any directory). Packages are only recognized if they include an __init__.py file." ([https://setuptools.readthedocs.io/en/latest/setuptools.html#using-find-packages](url)).

Adding empty __init__.py in those packages solved the problem for me.

### PR DESCRIPTION
**Pull request by edouardoyallon**
It takes of the doc with a little update of the website.

This pull requests updates the examples for the 3D end. All examples are in Pytorch currently.

Please review PR #432 and PR #434 before reviewing this pull request.

Close #309
Close #111
Close #421
Close #480

### CODE REVIEW

**Comment by edouardoyallon** [comment]
In order to avoid ill-conditionning warning, I originally removed one value we cross validate. But actually, I believe this is fine.(it's pretty neat from the output what's going on and one can guess that the corresponding value isn't good)

Does it sound good if we maintain it?(output below)

```
Computing solid harmonic scattering coefficients of 7165 molecules of QM7 database on GPU
sigma: 2.0, L: 3, J: 2, integral powers: [0.5, 1.0, 2.0, 3.0]
lseration 452 ETA: [02:05:34]:46]
Ridge regression, alpha: 1.0, MAE: 5.888091087341309, RMSE: 8.201009750366211
/users/data/oyallon/anaconda3/envs/torch/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.75391e-08): result may not be accurate.
  overwrite_a=True).T
/users/data/oyallon/anaconda3/envs/torch/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.98281e-08): result may not be accurate.
  overwrite_a=True).T
/users/data/oyallon/anaconda3/envs/torch/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.44838e-08): result may not be accurate.
  overwrite_a=True).T
/users/data/oyallon/anaconda3/envs/torch/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=2.05205e-08): result may not be accurate.
  overwrite_a=True).T
/users/data/oyallon/anaconda3/envs/torch/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=2.57147e-08): result may not be accurate.
  overwrite_a=True).T
Ridge regression, alpha: 0.1, MAE: 4.926825046539307, RMSE: 6.922209739685059
Ridge regression, alpha: 0.01, MAE: 4.043473720550537, RMSE: 5.792051315307617
Ridge regression, alpha: 0.001, MAE: 3.3518476486206055, RMSE: 4.858438968658447
Ridge regression, alpha: 0.0001, MAE: 2.96313738822937, RMSE: 4.426484107971191
Ridge regression, alpha: 1e-05, MAE: 2.757580280303955, RMSE: 4.213638782501221
Ridge regression, alpha: 1e-06, MAE: 2.669837713241577, RMSE: 4.095427989959717
Ridge regression, alpha: 1e-07, MAE: 2.6678569316864014, RMSE: 4.086413860321045
Ridge regression, alpha: 1e-08, MAE: 2.7660093307495117, RMSE: 4.100221157073975
Ridge regression, alpha: 1e-09, MAE: 2.8590779304504395, RMSE: 4.092794895172119
```

File: README.md [hunk 1]
```diff
@@ -17,14 +17,15 @@ Use Kymatio if you need a library that:
 * runs seamlessly on CPU and GPU hardware, with major deep learning APIs, such
   as PyTorch and TensorFlow.
 
+# The environment Kymatio
```
Comment (janden) [inline README.md]: “Kymatio environment”

File: README.md [hunk 2]
```diff
@@ -101,7 +102,24 @@ from kymatio.tensorflow import Scattering2D
...
+
+```
+pip install -r requirements.txt
+python setup.py develop
+```
+
+
+## GPU acceleration
 
 The available backends are PyTorch (`torch`), PyTorch+scikit-cuda (`torch_skcuda`), TensorFlow (`tensorflow`), and NumPy
```
Comment (janden) [inline README.md]: Three of these are frontends, no? Only `torch_skcuda` is a backend.

File: README.md [hunk 3]
```diff
@@ -170,7 +172,7 @@ cd doc; make clean; make html
 
 We wish to thank the Scientific Computing Core at the Flatiron Institute for the use of their computing resources for testing.
 
-[![Flatiron](doc/source/_static/FL_Full_Logo_Mark_Small.png)](https://www.simonsfoundation.org/flatiron)
+[![Flatiron](../../doc/source/_static/FL_Full_Logo_Mark_Small.png)](https://www.simonsfoundation.org/flatiron)
```
Comment (janden) [inline README.md]: `https://www.flatironinstitute.org/`

File: doc/source/conf.py [hunk 4]
```diff
@@ -19,14 +19,14 @@
 
 
 
-sys.path.insert(0, os.path.abspath('../../kymatio'))
+sys.path.insert(0, os.path.abspath('../..')) #/kymatio'))
```
Comment (janden) [inline doc/source/conf.py]: Remove comment. It's too risky to leave it in.

File: doc/source/conf.py [hunk 5]
```diff
@@ -19,14 +19,14 @@
...
 
 
 
 # -- Project information -----------------------------------------------------
 
 project = 'kymatio'
-copyright = '2018-2019, The Kymatio Developers'
+copyright = '2018-2020, The Kymatio Developers'
```
Comment (janden) [inline doc/source/conf.py]: En dash “–”, not hyphen “-”.

File: doc/source/conf.py [hunk 6]
```diff
@@ -81,8 +82,8 @@
 # The suffix(es) of source filenames.
 # You can specify multiple suffix as a list of string:
 #
-# source_suffix = ['.rst', '.md']
-source_suffix = '.rst'
+source_suffix = ['.rst', '.md']
+#source_suffix = '.rst'
```
Comment (janden) [inline doc/source/conf.py]: Remove.

File: doc/source/conf.py [hunk 7]
```diff
@@ -143,7 +144,16 @@
 # default: ``['localtoc.html', 'relations.html', 'sourcelink.html',
 # 'searchbox.html']``.
 #
-# html_sidebars = {}
+
+#html_sidebars = {
```
Comment (janden) [inline doc/source/conf.py]: Remove?

File: doc/source/developerguide.rst [hunk 8]
```diff
@@ -3,38 +3,67 @@
 Information for developers
 **************************
 
-Kymatio implements the scattering transform for different frontends (currently ``numpy``, ``torch``, ``tensorflow``), each of
-which have one or more corresponding backends. This way, the generic scattering algorithm can be written in an architecture-agnostic manner,
-since all low-level operations are relegated to the backend, and high-level operations specific to an API are relegated
-to the frontend.
+(GitHub Workflow)
```
Comment (janden) [inline doc/source/developerguide.rst]: TODO?

File: doc/source/developerguide.rst [hunk 9]
```diff
@@ -3,38 +3,67 @@
...
 
-Backend: core of the algorithm
-==============================
+Kymatio implements the scattering transform for different frontends (currently ``numpy``, ``torch``, ``tensorflow``),
+each of which have one or more corresponding backends. This way, the generic scattering algorithm can be written in an
+architecture-agnostic manner, since all low-level operations are relegated to the backend, and high-level operations
+specific to an API are relegated to the frontend.
 
-Common to the 1D, 2D and 3D scattering transform routine are four low-level functions which must be optimized:
+To make sure that a future Pull Request (PR) will pass the `jenkins` and `travis` tests, please try our package on the
```
Comment (janden) [inline doc/source/developerguide.rst]: Have “pull request” in lowercase, write out Jenkins and Travis in regular font.

File: doc/source/developerguide.rst [hunk 10]
```diff
@@ -3,38 +3,67 @@
...
+unit tests, the speed as well as the documentation. You might be willing to install auxiliary libraries via the
```
Comment (janden) [inline doc/source/developerguide.rst]: “be willing” → “need”

File: doc/source/developerguide.rst [hunk 11]
```diff
@@ -3,38 +3,67 @@
...
+``requirements_optional.txt``.
 
-1. Fast Fourier transform (FFT) and its inverse (IFFT)
+For development purposes, you might be willing to install the package via::
```
Comment (janden) [inline doc/source/developerguide.rst]: “be willing” → “need”

File: doc/source/developerguide.rst [hunk 12]
```diff
@@ -3,38 +3,67 @@
...
-If all the tests pass, you may submit your pull request as explained below.
+If all the tests pass, you may submit your pull request as explained below. A speed-test is welcome as well.
 
 Checking speed
 ==============
 
-Please check out the `examples/*d/compute_speed.py` scripts to benchmark your modifications.
+For checking the speed of a given PR, run the ASV benchmarks on various architectures via::
+
+    run asv
```
Comment (janden) [inline doc/source/developerguide.rst]: Might have to be in the benchmarks directory for this to work. Check.

File: doc/source/developerguide.rst [hunk 13]
```diff
@@ -3,38 +3,67 @@
...
+
+Thank you for reporting us your numbers!
```
Comment (janden) [inline doc/source/developerguide.rst]: Do we need this?

File: doc/source/developerguide.rst [hunk 14]
```diff
@@ -3,38 +3,67 @@
...
+
+Checking documentation
+======================
+
+For checking the documentation, please run the following commands, that will built it through sphinx::
+
+    make clean
```
Comment (janden) [inline doc/source/developerguide.rst]: Again, `cd doc` is needed here?

File: doc/source/developerguide.rst [hunk 15]
```diff
@@ -3,38 +3,67 @@
...
+    make html
 
 Proposing a pull request
 ========================
 
-Each pull request (PR) must be documented using docstrings, illustrated with an example and must pass the unit tests.
+Each PR must be documented using docstrings, illustrated with an example and must pass the unit tests. Please check the
+PRs already merged on the GitHub repository if you need an example of a good PR.
```
Comment (janden) [inline doc/source/developerguide.rst]: Yeah not sure all of these are good examples. Maybe we should have a template?

File: doc/source/index.rst [hunk 16]
```diff
@@ -1,80 +1,13 @@
+.. mdinclude:: ../../README.md
```
Comment (janden) [inline doc/source/index.rst]: :+1:

File: examples/3d/scattering3d_qm7_torch.py [hunk 17]
```diff
@@ -123,9 +124,7 @@ def get_qm7_positions_and_charges(sigma, overlapping_precision=1e-1):
     delta = sigma * np.sqrt(-8 * np.log(overlapping_precision))
     positions = positions * delta / min_dist
 
-    return (torch.from_numpy(positions),
-            torch.from_numpy(charges),
-            torch.from_numpy(valence_charges))
+    return positions, charges,valence_charges
```
Comment (janden) [inline examples/3d/scattering3d_qm7_torch.py]: Space after `,`.

File: examples/3d/scattering3d_qm7_torch.py [hunk 18]
```diff
@@ -159,28 +158,22 @@ def compute_qm7_solid_harmonic_scattering_coefficients(
...
-    grid = torch.from_numpy(
-        np.fft.ifftshift(
-            np.mgrid[-M//2:-M//2+M, -N//2:-N//2+N, -O//2:-O//2+O].astype('float32'),
-            axes=(1, 2, 3)))
+
+    if torch.cuda.is_available():
+        device = 'cuda'
+    else:
+        device = 'cpu'
+    grid =  np.fft.ifftshift( np.mgrid[-M//2:-M//2+M, -N//2:-N//2+N, -O//2:-O//2+O].astype('float32'),axes=(1, 2, 3))
```
Comment (janden) [inline examples/3d/scattering3d_qm7_torch.py]: No space before `np`. Space after `,`.

File: examples/3d/scattering3d_qm7_torch.py [hunk 19]
```diff
@@ -159,28 +158,22 @@ def compute_qm7_solid_harmonic_scattering_coefficients(
...
 
-    scattering = HarmonicScattering3D(J=J, shape=(M, N, O), L=L, sigma_0=sigma)
-
-    if cuda:
-        grid = grid.cuda()
-        pos = pos.cuda()
-        full_charges = full_charges.cuda()
-        valence_charges = valence_charges.cuda()
-        scattering.cuda()
+    scattering = HarmonicScattering3D(J=J, shape=(M, N, O), L=L, sigma_0=sigma, frontend='torch').to(device)
```
Comment (janden) [inline examples/3d/scattering3d_qm7_torch.py]: Change this to use `from kymatio.torch import HarmonicScattering3D` instead.

File: examples/3d/scattering3d_qm7_torch.py [hunk 20]
```diff
@@ -159,28 +158,22 @@ def compute_qm7_solid_harmonic_scattering_coefficients(
...
 
     order_0, orders_1_and_2 = [], []
     print('Computing solid harmonic scattering coefficients of {} molecules '
-        'of QM7 database on {}'.format(pos.size(0), 'GPU' if cuda else 'CPU'))
+        'of QM7 database on {}'.format(pos.shape[0], 'GPU' if device=='cuda' else 'CPU'))
```
Comment (janden) [inline examples/3d/scattering3d_qm7_torch.py]: Spaces around `==`.

File: kymatio/scattering3d/core/scattering3d.py [hunk 21]
```diff
@@ -1,7 +1,7 @@
 # Authors: Louis Thiry, Georgios Exarchakis
 # Scientific Ancestry: Louis Thiry, Georgios Exarchakis, Matthew Hirn, Michael Eickenberg
 
-def scattering3d(x, filters, rotation_covariant, L, J, max_order, backend, averaging):
+def scattering3d(x, filters, rotation_covariant, L, J, max_order, backend, averaging, vectorize=True):
```
Comment (janden) [inline kymatio/scattering3d/core/scattering3d.py]: Leftover from other PR? I can remove it.

File: requirements_optional.txt [hunk 22]
```diff
@@ -5,4 +5,6 @@ sphinx-gallery
 pillow
 matplotlib
 scikit-learn
-texext
\ No newline at end of file
+texext
+m2r
+
```
Comment (janden) [inline requirements_optional.txt]: No blank line.

File: doc/source/conf.py (hunk 5 above)
Comment (edouardoyallon) [reply doc/source/conf.py]: (I think it's good now!)

File: kymatio/scattering3d/core/scattering3d.py (hunk 21 above)
Comment (edouardoyallon) [reply kymatio/scattering3d/core/scattering3d.py]: Well, that's fine.

File: doc/source/developerguide.rst (hunk 15 above)
Comment (edouardoyallon) [reply doc/source/developerguide.rst]: We should .. I'm ok with them checking with what we did

File: doc/source/developerguide.rst (hunk 8 above)
Comment (edouardoyallon) [reply doc/source/developerguide.rst]: (subtitle, is it fine?)

File: kymatio/scattering3d/core/scattering3d.py (hunk 21 above)
Comment (edouardoyallon) [reply kymatio/scattering3d/core/scattering3d.py]: Wow gosh, sorry, I didn't catch what you meant

**Pull request merged**